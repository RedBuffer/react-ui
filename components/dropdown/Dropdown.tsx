import * as React from 'react';
import classnames from 'classnames';

export interface PropTypes {
  style?: React.CSSProperties;

  className?: string;

  children: React.ReactNode;

  overlay: React.ReactElement;

  trigger?: 'hover' | 'click';
}

const Dropdown = ({
  style,
  overlay,
  children,
  trigger = 'hover',
  className,
  ...restProps
}: PropTypes): React.ReactElement => {
  const [hidden, setHidden] = React.useState(true);

  return (
    <div
      className={classnames(
        'react-ui-dropdown',
        'm-4',
        'relative',
        'inline-block',
        className,
      )}
      style={style}
      onMouseEnter={() => (trigger === 'hover' ? setHidden(false) : {})}
      onMouseLeave={() => (trigger === 'hover' ? setHidden(true) : {})}
      onClick={() => (trigger === 'click' ? setHidden(!hidden) : {})}
      {...restProps}
    >
      {children}
      {!hidden ? (
        <div
          className={classnames(
            'react-ui-dropdown-content',
            'w-32',
            'absolute',
            'z-10',
          )}
        >
          {overlay}
        </div>
      ) : (
        <></>
      )}
    </div>
  );
};

export default Dropdown;
