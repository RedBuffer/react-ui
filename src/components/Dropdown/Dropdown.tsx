import * as React from 'react';
import classnames from 'classnames';

type Align = {
  points?: [string, string];
  offset?: [number, number];
  targetOffset?: [number, number];
  overflow?: {
    adjustX?: boolean;
    adjustY?: boolean;
  };
  useCssRight?: boolean;
  useCssBottom?: boolean;
  useCssTransform?: boolean;
};

interface Classes {
  dropdown?: string;

  button?: string;

  content?: string;
}

export interface PropTypes {
  style?: React.CSSProperties;

  className?: string;

  classes?: Classes;

  item: React.ReactElement;

  disabled?: boolean;

  align?: Align;
}

const Dropdown = ({
  style,
  item,
  disabled,
  className,
  classes = {},
  ...restProps
}: PropTypes): React.ReactElement => {
  return (
    <div
      className={classnames(
        'react-ui-dropdown',
        'm-4',
        'relative',
        'inline-block',
        className,
        classes.dropdown,
      )}
      {...restProps}
    >
      <button
        className={classnames(
          'react-ui-dropdown-btn',
          'p-4',
          'text-white',
          'bg-blue-400',
          'border-0',
          'cursor-pointer',
          classes.button,
        )}
        disabled={disabled}
      >
        Dropdown
      </button>
      <div
        className={classnames(
          'react-ui-dropdown-content',
          'hidden',
          'w-32',
          'absolute',
          'z-10',
          classes.content,
        )}
      >
        <a
          className="p-4 block no-underline border-gray-400 bg-gray-100 text-black"
          href="#"
        >
          Item 1
        </a>
        <a
          className="p-4 block no-underline border-gray-400 bg-gray-100 text-black"
          href="#"
        >
          Item 2
        </a>
        <a
          className="p-4 block no-underline border-gray-400 bg-gray-100 text-black"
          href="#"
        >
          Item 3
        </a>
      </div>
      <style jsx>
        {`
          .react-ui-dropdown-content a:hover {
            background-color: #f1f1f1;
          }
          .react-ui-dropdown:hover .react-ui-dropdown-content {
            display: block;
            box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
          }
        `}
      </style>
    </div>
  );
};

export default Dropdown;
