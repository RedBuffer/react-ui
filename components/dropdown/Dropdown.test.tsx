import * as React from 'react';
import { shallow, mount } from 'enzyme';

import Dropdown from './Dropdown';

describe('Dropdown', () => {
  let handleClick;
  let wrapper;
  const callHandleEdit = (item) => (e) => {
    e.stopPropagation();
    handleClick(item);
  };

  beforeEach(() => {
    handleClick = jest.fn((item) => item);
    wrapper = mount(
      <Dropdown
        overlay={
          <ul>
            <li>
              <a id="item-1" onClick={callHandleEdit('item 1')}>
                Item 1
              </a>
            </li>
            <li>
              <a id="item-2" onClick={callHandleEdit('item 2')}>
                Item 2
              </a>
            </li>
            <li>
              <a id="item-3" onClick={callHandleEdit('item 3')}>
                Item 3
              </a>
            </li>
          </ul>
        }
      />,
    );
  });

  it('should render correctly', () => {
    const component = shallow(<Dropdown overlay={<></>} />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly with menu', () => {
    const component = wrapper;

    expect(component).toMatchSnapshot();
  });

  it('click on menu item 1 ', () => {
    wrapper.find('#item-1').hostNodes().simulate('click');

    expect(handleClick.mock.calls[0][0]).toBe('item 1');
  });
});
