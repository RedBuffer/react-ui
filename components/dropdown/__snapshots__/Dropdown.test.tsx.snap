// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render correctly 1`] = `
<div
  className="react-ui-dropdown m-4 relative inline-block"
>
  <button
    className="react-ui-dropdown-btn p-4 text-white btn-primary border-0 cursor-pointer"
  >
    Dropdown
  </button>
  <div
    className="react-ui-dropdown-content hidden w-32 absolute z-10"
  />
</div>
`;

exports[`Dropdown should render correctly with menu 1`] = `
<Dropdown
  overlay={
    <ul>
      <li>
        <a
          id="item-1"
          onClick={[Function]}
        >
          Item 1
        </a>
      </li>
      <li>
        <a
          id="item-2"
          onClick={[Function]}
        >
          Item 2
        </a>
      </li>
      <li>
        <a
          id="item-3"
          onClick={[Function]}
        >
          Item 3
        </a>
      </li>
    </ul>
  }
>
  <div
    className="react-ui-dropdown m-4 relative inline-block"
  >
    <button
      className="react-ui-dropdown-btn p-4 text-white btn-primary border-0 cursor-pointer"
    >
      Dropdown
    </button>
    <div
      className="react-ui-dropdown-content hidden w-32 absolute z-10"
    >
      <ul>
        <li>
          <a
            id="item-1"
            onClick={[Function]}
          >
            Item 1
          </a>
        </li>
        <li>
          <a
            id="item-2"
            onClick={[Function]}
          >
            Item 2
          </a>
        </li>
        <li>
          <a
            id="item-3"
            onClick={[Function]}
          >
            Item 3
          </a>
        </li>
      </ul>
    </div>
  </div>
</Dropdown>
`;
